lib/src/main/scala/spinal/lib/tools/HDElkDiagramGen.scala

```diff
@@ -0,0 +1,599 @@
+              fanOut).getClass.getSimpleName
+        } else if (inIsBus == 1) {
+          if (!clkMap.contains(net.getName()))
+            newEdge.label = net.getName()
+        } else if (outIsBus == 1) {
+          if (!clkMap.contains(fanOut.getName()))
+            newEdge.label = fanOut.getName()
+        }
+        var isContained = false
+        for (thisEdge <- edges) {
+          if (thisEdge.source == newEdge.source && thisEdge.target == newEdge.target && thisEdge.label == newEdge.label)
+            isContained = true
+        }
+        if (!netIsWrong && !fanOutIsWrong && !isContained && newEdge.source != "" && newEdge.target != "")
+          edges.add(newEdge)
+      }
+    }
+  }
+
+  /** Generating HDElk language for Node */
+  private def drawNodes(thisNode: ElkNode): Unit = {
+    writer.write(s"""{id:"${thisNode.labelName}",\n""")
+    if (thisNode.typeName != "")
+      writer.write(s"""type:"${thisNode.typeName}",\n""")
+    if (thisNode.highlight != 100)
+      writer.write(s"""highlight:${thisNode.highlight},\n""")
+    if (thisNode.inPorts.nonEmpty) {
+      writer.write(s"""inPorts: [""")
+      for (inPort <- thisNode.inPorts) {
+        if (inPort.highlight != 1)
+          writer.write(
+            s"""{id:"${inPort.name}",highlight:${inPort.highlight}},""")
+        else writer.write(s""""${inPort.name}",""")
+      }
+      writer.write(s"""],\n""")
+    }
+    if (thisNode.outPorts.nonEmpty) {
+      writer.write(s"""outPorts: [""")
+      for (outPort <- thisNode.outPorts) {
+        if (outPort.highlight != 1)
+          writer.write(
+            s"""{id:"${outPort.name}",highlight:${outPort.highlight}},""")
+        else writer.write(s""""${outPort.name}",""")
+      }
+      writer.write(s"""],\n""")
+    }
+    if (thisNode.children.nonEmpty) {
+      writer.write(s"""children: [\n""")
+      for (thisChildren <- thisNode.children) drawNodes(thisChildren)
+      writer.write(s"""],\n""")
+    }
+    if (thisNode == topNode) drawEdges()
+    writer.write(s"""},\n""")
+  }
+
+  /** Generating HDElk language for Edge */
+  private def drawEdges(): Unit = {
+    writer.write(s"""edges:[\n""")
+    for (edge <- edges) {
+      writer.write(
+        s"""{ source:"${edge.source}",target:"${edge.target}",bus:${edge.isBus},""")
+      if (edge.label != "")
+        writer.write(s"""label:"${edge.label}",""")
+      if (edge.highlight != 100)
+        writer.write(s"""highlight:${edge.highlight}""")
+      writer.write(s"""},\n""")
+    }
+    writer.write(s"""]\n""")
+  }
+
+  /** Generating HDElk language for Clock legend */
+  private def drawClockDomains(): Unit = {
+    writer.write(s"""{id:"ClockDomains",\nchildren:[\n""")
+    for (element <- clkNamesMap) {
+      writer.write(s"""{id:"${element._1}",highlight:${element._2}},\n""")
+    }
+    writer.write(s"""]\n}\n""")
+  }
+
+  /** Integrating all methods to draw an image */
+  def beginDraw(): Unit = {
```

Comment:
You should pass the writer as a parameter instead of opening it for append during construction and closing it here.